### PR TITLE
feat: staking payout refactor

### PR DESCRIPTION
**Pull request by wischli**
### Follow-up PRs
* #413 
* #414

### TLDR Description
* Refactors staking payouts to not happen automatically
  * Instead, a counter (similar to the Substrate staking pallet participation points) is incremented for each authored block
  * The counter can be consumed and converted into rewards
* Adds extrinsic `claim_rewards` which enables any signed origin to trigger a payout for themselves

### Lengthy Description

The proposed solution goes more into the direction of Substrate staking. Every time a collator authors a block, their rewards counter is incremented. The counter can be consumed to increment accumulated rewards. The rewards need to be actively claimed by the new `claim_rewards_for` extrinsic. 

There are a lot of events which automatically consume the `RewardCounter` and increment the accumulated rewards since the last actively requested payout:

- Collator changes stake (more, less)
- A delegator changes stake (more, less)
- A delegator removes delegation
- A delegator is replaced
- A collator leaves
- The inflation configuration changes

### Background

The main reason for this refactor is that the current solution is not highly scalable as it inherently emits a `Rewarded` event for each address that gets rewarded. For the current configuration of delegators, this is rather unproblematic. 

Moreover, we currently enforce the taxable `Rewarded` event onto all stakers and it occurs very frequently. Therefore, it is quite a challenge to summarize all such events.

### TODO

- [x] More thorough description
- [x] Add claiming extrinsic
- [x] Refactor yearly inflation adjustment
- [x] Unit tests
- [x] Benchmarks

#### Extrinsics 
- [x] set_inflation
- [x] force_remove_candidate
- [x] init_leave_candidates
- [x] candidate_stake_more
- [x] candidate_stake_less
- [x] join_delegators
- [x] delegate_another_candidate
- [x] leave_delegators
- [x] revoke_delegation
- [x] delegator_stake_more
- [x] delegator_stake_less

## Checklist:

- [x] I have verified that the code works
  - [x] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

File: pallets/parachain-staking/src/lib.rs [hunk 1]
```diff
@@ -2014,6 +2052,25 @@ pub mod pallet {
 
 			Ok(Some(<T as pallet::Config>::WeightInfo::unlock_unstaked(unstaking_len)).into())
 		}
+
+		// TODO: Benchmark, unit tests, docs
+		#[pallet::weight(0)]
+		pub fn claim_rewards_for(origin: OriginFor<T>, target: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/lib.rs:2058]: Any specific reason why we would allow anyone to trigger a payout for anyone else? I still don't understand what would be the use case for this, if I want to have control over when I would claim my rewards. If I want to delegate, I could set up a proxy,

File: pallets/parachain-staking/src/lib.rs [hunk 2]
```diff
@@ -2790,78 +2924,22 @@ pub mod pallet {
...
-			if let Some(state) = CandidatePool::<T>::get(author.clone()) {
-				let total_issuance = T::Currency::total_issuance();
-				let TotalStake {
-					collators: total_collators,
-					delegators: total_delegators,
-				} = <TotalCollatorStake<T>>::get();
-				let c_staking_rate = Perquintill::from_rational(total_collators, total_issuance);
-				let d_staking_rate = Perquintill::from_rational(total_delegators, total_issuance);
-				let inflation_config = <InflationConfig<T>>::get();
+			if CandidatePool::<T>::get(author.clone()).is_some() {
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/lib.rs:2933]: Probably unrelated, but we can use references to access the storage instead of cloning the key. Like `CandidatePool::<T>::get(&author)`.

**Review by ntn-x2 — COMMENTED** [review]
It will probably look much easier to follow and understand once refactored, so we are on the right path 🙂

**Review by ntn-x2 — COMMENTED** [review]
I have one main question about the rewards for delegators. It seems to me they are incentivised to jump onboard right before some other event is supposed to happen (like a stake change or an inflation change). Is my understanding correct? Is that by design? Or am I missing something?

File: pallets/parachain-staking/src/lib.rs [hunk 3]
```diff
@@ -2773,95 +2383,121 @@ pub mod pallet {
...
+			let del_reward_count = RewardCount::<T>::get(acc);
+			let col_reward_count = RewardCount::<T>::get(col);
+			let diff = col_reward_count.saturating_sub(del_reward_count);
+
+			// only update if collator has higher reward count
+			if diff > 0 {
+				Rewards::<T>::mutate(acc, |reward| {
+					*reward = reward.saturating_add(Self::calc_block_rewards_delegator(stake, diff.into()));
+				});
+				RewardCount::<T>::insert(acc, col_reward_count);
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:2472]: When you request a payout as a delegator, the number of rewards you receive is dependent on your counter and the one of your collator. A delegator's counter can at most be even with the corresponding collator counter. When a delegator increments their rewards, the counter is equalized is aligned with the collator one to reflect that the rewards can only be incremented further after the collator authored at least another block.

_Example: 
1. Collator Alice has counter 100 and Bob joins as a delegator. Since he should only receive rewards from now on, his counter is initialized with 100. 
2. When Alice's counter is at 200, Bob stakes more and triggers a reward increment. He receives `(200-100)*reward_per_block(Bob)`. He is then treated as a newcomer, thus "resetting" his counter to 200._

File: pallets/parachain-staking/src/lib.rs [hunk 4]
```diff
@@ -1569,163 +1472,8 @@ pub mod pallet {
...
-
-			// Update states
-			CandidatePool::<T>::insert(&collator, state);
-			DelegatorState::<T>::insert(&acc, delegator);
-			<LastDelegation<T>>::insert(&acc, delegation_counter);
-
-			// update or clear storage of potentially kicked delegator
-			Self::update_kicked_delegator_storage(maybe_kicked_delegator);
+			// initiate reward counter to match the current state of the candidate
+			RewardCount::<T>::insert(&acc, RewardCount::<T>::get(&collator));
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:1476]: As you pointed out below, the description of `RewardCount` is incorrect. In the early design, I hoped to only require oit for collators but realized it is also needed for delegators.

File: pallets/parachain-staking/src/lib.rs [hunk 5]
```diff
@@ -2014,26 +1697,154 @@ pub mod pallet {
...
+			Ok(())
+		}
+
+		/// Executes the annual reduction of the reward rates for collators and
+		/// delegators. Moreover, sets rewards for all collators and delegators
+		/// before adjusting the inflation.
+		///
+		/// Emits `RoundInflationSet`.
+		#[pallet::weight(<T as Config>::WeightInfo::execute_pending_reward_change(T::MaxTopCandidates::get(), T::MaxDelegatorsPerCollator::get()))]
+		pub fn execute_pending_reward_change(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:1782]: I agree. Either of both options is better. Personally, I prefer _scheduled_.

File: pallets/parachain-staking/src/lib.rs (hunk 1 above)
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:1717]: IMO, the main reason would be to cleanup storage and consume tiny rewards that have not been claimed by potentially exited users. I understand that's a very unlikely scenario and each user has an incentive to claim their rewards.

Moreover, in frame staking this is also possible by triggering a payout for a validator which rewards their nominators as well.

Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:1717]: I am completely fine changing this and would be interested in @weichweich 's opinion.

Comment (ntn-x2) [reply pallets/parachain-staking/src/lib.rs:1717]: Yeah I think Parity does that because otherwise users would just forget and keep piling up stuff, I think. Which would be fine in our case if that is the assumption. But of course the downside is that I can't simply decide to trigger my reward at any time, because someone else inside my delegation pool might do it for me.

File: pallets/parachain-staking/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:2472]: I will add a comment to explain why it is not reset to zero.

[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/parachain-staking/src/lib.rs:2472]: Ok, understood. Thanks.

[the remark this answers is not in the document]
Comment (ntn-x2) [reply pallets/parachain-staking/src/lib.rs:2472]: I guess it's just a design decision, whether to substract the collator count from the total, or sum the delegator count to the collator count. I would have gone the second road as it makes it easier to understand, but yeah a comment explaining that would help.

File: pallets/parachain-staking/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: [ad13650](https://github.com/KILTprotocol/kilt-node/pull/384/commits/ad13650e92a6c87335b1ac3374bfc03531c6025f)

File: pallets/parachain-staking/src/types.rs [hunk 6]
```diff
@@ -211,99 +209,60 @@ where
...
-				self.delegations
-					.mutate(|vec| vec[i].amount = vec[i].amount.saturating_sub(less));
-				self.total = self.total.saturating_sub(less);
-				self.delegations.sort_greatest_to_lowest();
-				Some(Some(self.delegations[i].amount))
+	/// Returns Ok(Some(delegated_amount)) if successful, `Err` if delegation
+	/// was not found and Ok(None) if delegated stake would underflow.
+	pub fn dec_delegation(&mut self, collator: AccountId, less: Balance) -> Result<Option<Balance>, ()> {
+		if self.owner == Some(collator) {
+			if self.amount > less {
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/types.rs]: [ad13650](https://github.com/KILTprotocol/kilt-node/pull/384/commits/ad13650e92a6c87335b1ac3374bfc03531c6025f)

File: pallets/parachain-staking/src/mock.rs [hunk 7]
```diff
@@ -370,6 +366,30 @@ pub(crate) fn roll_to(n: BlockNumber, authors: Vec<Option<AccountId>>) {
 	}
 }
 
+#[allow(unused_must_use)]
+pub(crate) fn roll_to_claim_rewards(n: BlockNumber, authors: Vec<Option<AccountId>>) {
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/mock.rs:387]: [ad13650](https://github.com/KILTprotocol/kilt-node/pull/384/commits/ad13650e92a6c87335b1ac3374bfc03531c6025f)

File: pallets/parachain-staking/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:2481]: [ad13650](https://github.com/KILTprotocol/kilt-node/pull/384/commits/ad13650e92a6c87335b1ac3374bfc03531c6025f)

File: pallets/parachain-staking/src/lib.rs [hunk 8]
```diff
@@ -664,6 +647,22 @@ pub mod pallet {
 	#[pallet::getter(fn last_reward_reduction)]
 	pub(crate) type LastRewardReduction<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
 
+	/// The counter of accumulated rewards for a collator candidate.
+	///
+	/// It maps from collator accounts to the number of block authored by them
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: [ad13650](https://github.com/KILTprotocol/kilt-node/pull/384/commits/ad13650e92a6c87335b1ac3374bfc03531c6025f)

File: pallets/parachain-staking/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:2481]: How would your proposal work? We basically don't want to increase the delegator counter ever, just when initializing/resetting, we need to align with the collator one in my case. I decided for that to decrease storage writes.

File: pallets/parachain-staking/src/lib.rs [hunk 9]
```diff
@@ -1778,8 +1785,8 @@ pub mod pallet {
 		/// before adjusting the inflation.
 		///
 		/// Emits `RoundInflationSet`.
-		#[pallet::weight(<T as Config>::WeightInfo::execute_pending_reward_change(T::MaxTopCandidates::get(), T::MaxDelegatorsPerCollator::get()))]
-		pub fn execute_pending_reward_change(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+		#[pallet::weight(<T as Config>::WeightInfo::exectue_scheduled_reward_change(T::MaxTopCandidates::get(), T::MaxDelegatorsPerCollator::get()))]
+		pub fn exectue_scheduled_reward_change(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/lib.rs:1789]:
```suggestion
		pub fn execute_scheduled_reward_change(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
```

File: pallets/parachain-staking/src/mock.rs [hunk 10]
```diff
@@ -355,6 +355,15 @@ pub(crate) fn almost_equal(left: Balance, right: Balance, precision: Perbill) ->
 	left.max(right) - left.min(right) <= err
 }
 
+/// Incrementelly traverses from the current block to the provided one and
+/// potentially sets a block authors.
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/mock.rs:359]:
```suggestion
/// potentially sets block authors.
```

File: pallets/parachain-staking/src/lib.rs (hunk 9 above)
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:1789]: Oh boy, shouldn't "quickly push before lunch". Thanks!

File: pallets/parachain-staking/src/lib.rs (hunk 1 above)
Comment (weichweich) [reply pallets/parachain-staking/src/lib.rs:1734]: Let's first start with the more restrictive version. We might be able to change that later if we feel the need for it.

Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:1734]: By more restrictive you are referring to Antonio's proposal, e.g. origins can only claim for themselves?

Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: [420170d](https://github.com/KILTprotocol/kilt-node/pull/384/commits/420170d387972e64c0e0c50997f8ca1110cac9f4)

**Comment by wischli** [comment]
@weichweich @ntn-x2 Could you review my additions of runtime API calls ([6f73296](https://github.com/KILTprotocol/kilt-node/pull/384/commits/6f732963fabdce3408d75bd73fb5394fd3e3a8cc),  [20fd523](https://github.com/KILTprotocol/kilt-node/pull/384/commits/20fd5231ff1a6bffd260aef53b9d06f43176c225) and [0f8160b](https://github.com/KILTprotocol/kilt-node/pull/384/commits/0f8160b18cb3a205824a805cc08cee2a339e69d1)) when you have the time. Thank you!

I also made claiming stricter. IMO, this PR is done now.

File: pallets/parachain-staking/src/lib.rs [hunk 11]
```diff
@@ -2733,95 +2352,176 @@ pub mod pallet {
...
+		pub fn get_unclaimed_staking_rewards(acc: &T::AccountId) -> BalanceOf<T> {
+			let mut reward_count = RewardCount::<T>::get(acc);
+			let rewards = Rewards::<T>::get(acc);
+
+			if let Some(delegator_state) = DelegatorState::<T>::get(acc) {
+				// delegator reward count does not automatically inrement in order to be
+				// scalable
+				reward_count =
+					reward_count.saturating_add(delegator_state.owner.map(RewardCount::<T>::get).unwrap_or(0u32));
+				let stake = DelegatorState::<T>::get(acc)
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/lib.rs:2403]: Don't we have already the `delegator_state` variable?

File: pallets/parachain-staking/src/lib.rs [hunk 12]
```diff
@@ -2733,95 +2352,176 @@ pub mod pallet {
...
+			}
+		}
+
+		/// Calculates the current staking and reward rates for collators and
+		/// delegators.
+		///
+		/// At least used in Runtime API.
+		pub fn get_staking_rates() -> runtime_api::StakingRates {
+			let total_issuance = T::Currency::total_issuance();
+			let total_stake = <TotalCollatorStake<T>>::get();
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/lib.rs:2484]:
```suggestion
			let total_stake = TotalCollatorStake::<T>::get();
```

File: pallets/parachain-staking/src/runtime_api.rs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/runtime_api.rs:18]: Any reason why it has not been put inside the `rpc` folder? I know this is not technically RPC, but it would help to keep the same structure over the codebase.

File: pallets/parachain-staking/src/runtime_api.rs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	pub collator_staking_rate: Perquintill,
+	pub collator_reward_rate: Perquintill,
+	pub delegator_staking_rate: Perquintill,
+	pub delegator_reward_rate: Perquintill,
+}
+
+sp_api::decl_runtime_apis! {
+	pub trait ParachainStakingApi<AccountId, Balance>
+	where
+		AccountId:  Eq + PartialEq + Debug + Encode + Decode + Clone,
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/runtime_api.rs:36]: I think we can get along with simply requiring `Codec`.

File: pallets/parachain-staking/src/runtime_api.rs [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+		Balance: Encode + Decode + MaxEncodedLen + Copy + Clone + Debug + Eq + PartialEq
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/runtime_api.rs:37]: Same here.

File: pallets/parachain-staking/src/tests.rs [hunk 16]
```diff
@@ -3926,3 +3836,504 @@ fn update_total_stake_no_collator_changes() {
...
+			);
+			assert_noop!(
+				StakePallet::increment_delegator_rewards(Origin::signed(1)),
+				Error::<Test>::DelegatorNotFound
+			);
+		});
+}
+
+#[test]
+fn api_get_unclaimed_staking_rewards() {
```
Comment (ntn-x2) [inline pallets/parachain-staking/src/tests.rs:4273]: Can we actually test that the returned value of the function actually matches what is unlocked when calling the extrinsic to pay out those rewards? What do you think?

File: pallets/parachain-staking/src/runtime_api.rs (hunk 14 above)
Comment (ntn-x2) [reply pallets/parachain-staking/src/runtime_api.rs:36]: At least that is the conclusion I got when working on the public credentials stuff.

**Review by ntn-x2 — COMMENTED** [review]
Few more comments, but it looks very close 🤗

File: pallets/parachain-staking/src/tests.rs (hunk 16 above)
Comment (wischli) [reply pallets/parachain-staking/src/tests.rs:4273]: [e703493](https://github.com/KILTprotocol/kilt-node/pull/384/commits/e703493c668c3eb2af734920ac05abca615de5c6)

File: pallets/parachain-staking/src/runtime_api.rs (hunk 13 above)
Comment (wischli) [reply pallets/parachain-staking/src/runtime_api.rs:18]: https://github.com/KILTprotocol/kilt-node/pull/413

File: pallets/parachain-staking/src/runtime_api.rs (hunk 14 above)
Comment (wischli) [reply pallets/parachain-staking/src/runtime_api.rs:36]: https://github.com/KILTprotocol/kilt-node/pull/413

File: pallets/parachain-staking/src/runtime_api.rs (hunk 15 above)
Comment (wischli) [reply pallets/parachain-staking/src/runtime_api.rs:37]: https://github.com/KILTprotocol/kilt-node/pull/413

File: pallets/parachain-staking/src/lib.rs [hunk 17]
```diff
@@ -2733,95 +2352,176 @@ pub mod pallet {
...
+		}
+
+		/// Calculates the staking rewards for a given account address.
+		///
+		/// At least used in Runtime API.
+		pub fn get_unclaimed_staking_rewards(acc: &T::AccountId) -> BalanceOf<T> {
+			let mut reward_count = RewardCount::<T>::get(acc);
+			let rewards = Rewards::<T>::get(acc);
+
+			if let Some(delegator_state) = DelegatorState::<T>::get(acc) {
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:2399]: [a3b2631](https://github.com/KILTprotocol/kilt-node/pull/384/commits/a3b263163aa3ee4772e51d9e460e914873a83386)

**Review by ntn-x2 — APPROVED** [review]
LGTM!

**Comment by wischli** [comment]
@weichweich Please review when you have the time, thanks!

**Review by weichweich — APPROVED** [review]
Looks good. I would say we can merge it as it is. I just have a few naming suggestions and the `RewardCounter` is a bit hard to understand. Maybe my suggestion could make this part a bit easier to understand?

The yearly inflation config update is moved from `on_initialize` to a call that can be called by anyone? Is there a reason for that?

File: pallets/parachain-staking/src/lib.rs [hunk 18]
```diff
@@ -2622,81 +2301,22 @@ pub mod pallet {
...
-				Self::deposit_event(Event::RoundInflationSet(
-					new_inflation.collator.max_rate,
-					new_inflation.collator.reward_rate.per_block,
-					new_inflation.delegator.max_rate,
-					new_inflation.delegator.reward_rate.per_block,
-				));
-				<T as Config>::WeightInfo::on_initialize_new_year();
-			}
-			T::DbWeight::get().reads(1)
-		}
```
[the remark this answers is not in the document]
Comment (weichweich) [reply pallets/parachain-staking/src/lib.rs:2678]: because it's too heavy we moved it to a call?

[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:2678]: This was moved to the `execute_scheduled_reward_change` function which can be called by any origin. However, this call [will fail when it's too early to be called](https://github.com/KILTprotocol/kilt-node/pull/384/files#diff-842251b834245800808bd740184c1e3b18c6af1f95fb1bd4d484420017fdf86eR1752-R1759).

The downside is, that the yearly adjustment has to be manually called and thus requires some kind of notification mechanism, e.g. a calendar entry. The upside is that it does not impose a risk exceeding the PoV size in inherent calls anymore.

File: pallets/parachain-staking/src/types.rs [hunk 19]
```diff
@@ -211,99 +209,58 @@ where
...
 	///
-	/// If already delegating to the same account, this call returns false and
-	/// doesn't insert the new delegation.
-	pub fn add_delegation(&mut self, stake: Stake<AccountId, Balance>) -> Result<bool, ()> {
-		let amt = stake.amount;
-		if self.delegations.try_insert(stake).map_err(|_| ())? {
-			self.total = self.total.saturating_add(amt);
-			Ok(true)
+	/// If already delegating to someone, this call will fail.
+	pub fn add_delegation(&mut self, stake: Stake<AccountId, Balance>) -> Result<(), ()> {
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/types.rs]: [30be731](https://github.com/KILTprotocol/kilt-node/pull/384/commits/30be731402178ef7ce10519459d1d228b8d5fec6)

File: pallets/parachain-staking/src/types.rs [hunk 20]
```diff
@@ -211,99 +209,58 @@ where
...
-#[scale_info(skip_type_params(MaxCollatorsPerDelegator))]
-#[codec(mel_bound(AccountId: MaxEncodedLen, Balance: MaxEncodedLen))]
-pub struct Delegator<AccountId: Eq + Ord, Balance: Eq + Ord, MaxCollatorsPerDelegator: Get<u32>> {
-	pub delegations: OrderedSet<Stake<AccountId, Balance>, MaxCollatorsPerDelegator>,
-	pub total: Balance,
-}
-
-impl<AccountId, Balance, MaxCollatorsPerDelegator> Delegator<AccountId, Balance, MaxCollatorsPerDelegator>
+pub type Delegator<AccountId, Balance> = Stake<Option<AccountId>, Balance>;
+impl<AccountId, Balance> Delegator<AccountId, Balance>
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/types.rs:213]: [30be731](https://github.com/KILTprotocol/kilt-node/pull/384/commits/30be731402178ef7ce10519459d1d228b8d5fec6)

File: pallets/parachain-staking/src/types.rs [hunk 21]
```diff
@@ -211,99 +209,58 @@ where
...
-			// amount is irrelevant for removal
-			amount: Balance::zero(),
-		});
-
-		if let Some(Stake::<AccountId, Balance> { amount: balance, .. }) = amt {
-			self.total = self.total.saturating_sub(balance);
-			Some(self.total)
+	/// Returns Ok if the delegation for the
+	/// collator exists and `Err` otherwise.
+	pub fn rm_delegation(&mut self, collator: AccountId) -> Result<(), ()> {
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/types.rs]: [30be731](https://github.com/KILTprotocol/kilt-node/pull/384/commits/30be731402178ef7ce10519459d1d228b8d5fec6)

File: pallets/parachain-staking/src/lib.rs [hunk 22]
```diff
@@ -1707,80 +1466,28 @@ pub mod pallet {
...
 		pub fn leave_delegators(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let acc = ensure_signed(origin)?;
 			let delegator = DelegatorState::<T>::get(&acc).ok_or(Error::<T>::DelegatorNotFound)?;
-			let num_delegations: u32 = delegator.delegations.len().saturated_into();
-			for stake in delegator.delegations.into_iter() {
-				Self::delegator_leaves_collator(acc.clone(), stake.owner.clone())?;
-			}
+			// should never throw
+			let collator = delegator.owner.ok_or(Error::<T>::DelegationNotFound)?;
+			Self::delegator_leaves_collator(acc.clone(), collator)?;
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: I moved the update of `InflationInfo` for the two extrinsics (`set_inflation` and `adjust_yearly_inflation_rate`) to a helper function and added [more documentation explaining why this can be called by any signed origin](https://github.com/KILTprotocol/kilt-node/pull/384/commits/f459050e76ebed0c6d142a70d13a9f66730d60e5#diff-842251b834245800808bd740184c1e3b18c6af1f95fb1bd4d484420017fdf86eR1727-R1729).

File: pallets/parachain-staking/src/lib.rs [hunk 23]
```diff
@@ -2733,95 +2353,179 @@ pub mod pallet {
...
+		/// blocks based on the given stake.
+		///
+		/// Depends on the current total issuance and staking reward
+		/// configuration for delegators.
+		fn calc_block_rewards_delegator(stake: BalanceOf<T>, multiplier: BalanceOf<T>) -> BalanceOf<T> {
+			let total_issuance = T::Currency::total_issuance();
+			let TotalStake {
+				delegators: total_delegators,
+				..
+			} = <TotalCollatorStake<T>>::get();
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: [cd9684e](https://github.com/KILTprotocol/kilt-node/pull/384/commits/cd9684e99508da4889354eb6d5db86c360c44311)

File: pallets/parachain-staking/src/lib.rs [hunk 24]
```diff
@@ -2733,95 +2353,179 @@ pub mod pallet {
...
+		/// many blocks based on the given stake.
+		///
+		/// Depends on the current total issuance and staking reward
+		/// configuration for collators.
+		fn calc_block_rewards_collator(stake: BalanceOf<T>, multiplier: BalanceOf<T>) -> BalanceOf<T> {
+			let total_issuance = T::Currency::total_issuance();
+			let TotalStake {
+				collators: total_collators,
+				..
+			} = <TotalCollatorStake<T>>::get();
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: [cd9684e](https://github.com/KILTprotocol/kilt-node/pull/384/commits/cd9684e99508da4889354eb6d5db86c360c44311)

File: pallets/parachain-staking/src/lib.rs [hunk 25]
```diff
@@ -1974,26 +1668,153 @@ pub mod pallet {
...
+		pub fn claim_rewards(origin: OriginFor<T>) -> DispatchResult {
+			let target = ensure_signed(origin)?;
+
+			// we could kill the storage entry but let's be safe in case the deposit fails
+			let rewards = Rewards::<T>::get(&target);
+			ensure!(!rewards.is_zero(), Error::<T>::RewardsNotFound);
+
+			// mint into target and reset rewards
+			let rewards = T::Currency::deposit_into_existing(&target, rewards)?;
+			Rewards::<T>::remove(&target);
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: Part of [f459050](https://github.com/KILTprotocol/kilt-node/pull/384/commits/f459050e76ebed0c6d142a70d13a9f66730d60e5)

File: pallets/parachain-staking/src/lib.rs (hunk 22 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: [55c6ac3](https://github.com/KILTprotocol/kilt-node/pull/384/commits/55c6ac30dd1771e0cd36ee6fa80e01475d912870)

File: pallets/parachain-staking/src/lib.rs [hunk 26]
```diff
@@ -624,6 +611,29 @@ pub mod pallet {
...
+	/// For collators, it reflects the number of authored blocks since the last
+	/// reward raise. Thus, everytime a collator authors a block, the
+	/// counter is increased. It is reset, when the collator increments their
+	/// rewards.
+	///
+	/// For delegators, it is used to determine the difference between the
+	/// delegator and corresponding collator when incrementing the delegator's
+	/// rewards. In this case, the counter is never incremented but reset to the
+	/// collator one when the delegator reward increment happens.
+	// TODO: Maybe rather use u64. Assuming 30 validators, u32 would suffice for 27 years of constant 12s blocktime.
```
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: Definitely much cleaner now but big refactor of the PR: [f459050](https://github.com/KILTprotocol/kilt-node/pull/384/commits/f459050e76ebed0c6d142a70d13a9f66730d60e5)

File: pallets/parachain-staking/src/lib.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs:2678]: I moved the update of `InflationInfo` for the two extrinsics (`set_inflation` and `adjust_yearly_inflation_rate`) to a helper function and added [more documentation explaining why this can be called by any signed origin](https://github.com/KILTprotocol/kilt-node/pull/384/commits/f459050e76ebed0c6d142a70d13a9f66730d60e5#diff-842251b834245800808bd740184c1e3b18c6af1f95fb1bd4d484420017fdf86eR1727-R1729).

File: pallets/parachain-staking/src/lib.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (wischli) [reply pallets/parachain-staking/src/lib.rs]: A nice side effect: We [do not have to increment delegator rewards anymore when incrementing collator rewards](https://github.com/KILTprotocol/kilt-node/pull/384/commits/f459050e76ebed0c6d142a70d13a9f66730d60e5#diff-842251b834245800808bd740184c1e3b18c6af1f95fb1bd4d484420017fdf86eL2434-L2450). This heavily decreases tx fees for collators, when they adjust their stake amounts. 

cc @ntn-x2

File: pallets/parachain-staking/src/lib.rs [hunk 27]
```diff
@@ -1527,10 +1424,11 @@ pub mod pallet {
 			// update states
 			CandidatePool::<T>::insert(&collator, state);
 			DelegatorState::<T>::insert(&acc, delegator_state);
-			<LastDelegation<T>>::insert(&acc, delegation_counter);
+			LastDelegation::<T>::insert(&acc, delegation_counter);
 
-			// update or clear storage of potentially kicked delegator
-			Self::update_kicked_delegator_storage(maybe_kicked_delegator);
+			// initiate rewarded counter to match the current autored counter of the
+			// candidate
```
Comment (weichweich) [inline pallets/parachain-staking/src/lib.rs:1430]:
```suggestion
			// initiate rewarded counter to match the current authored counter of the
			// candidate
```

**Review by weichweich — APPROVED** [review]
LGTM very beautiful PR. Love the links to the commits. Nice comments.
